### PR TITLE
fix #274965: "Save As" should work as it works in every other program.

### PR DESCRIPTION
**Pull request by mattmcclinch**
See https://musescore.org/en/node/274965.

### CODE REVIEW

**Comment by mattmcclinch** [comment]
This takes care of "Save a Copy" as well as "Save As", but "Save Selection" and "Export" are handled elsewhere in the code.

**Comment by mattmcclinch** [comment]
Now these dialogs are affected by this PR:
- "Save As..."
- "Save a Copy..."
- "Save Selection..."
- "Export..."

As you say, "Open..." can probably stay as it is.

**Comment by MarcSabatella** [comment]
How about Export Parts?

**Comment by mattmcclinch** [comment]
Now "Export Parts..." is affected also.

File: mscore/file.cpp [hunk 1]
```diff
@@ -1618,6 +1618,8 @@ void MuseScore::exportFile()
       if (lastSaveDirectory.isEmpty())
             lastSaveDirectory = settings.value("lastSaveDirectory", preferences.getString(PREF_APP_PATHS_MYSCORES)).toString();
       QString saveDirectory = lastSaveCopyDirectory;
+      if (cs->masterScore()->fileInfo()->exists())
+            saveDirectory = cs->masterScore()->fileInfo()->dir().path();
```
Comment (IsaacWeiss) [inline mscore/file.cpp]: Doesn't this always initialize `saveDirectory` before checking for the current score path? Wouldn't something like

`QString saveDirectory = (cs->masterScore()->fileInfo()->exists() ? cs->masterScore()->fileInfo()->dir().path() : lastSaveCopyDirectory);`

be simpler?

Comment (mattmcclinch) [reply mscore/file.cpp]: Yes it does. Would it be simpler? Probably. The line for "Save As" would become
```
      QString saveDirectory = cs->masterScore()->fileInfo()->exists() ? cs->masterScore()->fileInfo()->dir().path() : saveCopy ? mscore->lastSaveCopyDirectory : mscore->lastSaveDirectory;
```
which is a little long, and maybe a little less clear.

Comment (mattmcclinch) [reply mscore/file.cpp]: I admit I was not considering the efficiency of the QString assignment. If it were simply a pointer or scalar assignment, that would be a different matter.

**Comment by mattmcclinch** [comment]
I reorganized the logic to avoid unnecessary lookups and assignments. It did not seem necessary for exportFile() and exportParts() to make sure that MuseScore::lastSaveDirectory is not empty, since they both only use MuseScore::lastSaveCopyDirectory, so I removed that bit. And I verified that the final check for whether or not saveDirectory.isEmpty() is indeed necessary.

**Comment by shoogle** [comment]
Since some people like the current behaviour, it might be worth creating a new [advanced preference](https://github.com/musescore/MuseScore/blob/fde1785333cda38ea001fe39dc9d61338f6e38f9/mscore/preferences.h#L82) to retain it. Maybe a [boolean preference](https://github.com/musescore/MuseScore/blob/fde1785333cda38ea001fe39dc9d61338f6e38f9/mscore/preferences.cpp#L76) along the lines of:

- `"file/saveLoad/dialogsRemembersLastLocation" ` (default value: `FALSE`)

or alternatively:

- `"file/saveLoad/dialogsStartAtCurrentScoreLocation" ` (default value: `TRUE`)

Maybe there should be different preferences for each dialog, and maybe the defaults should be different for the export dialogs, because while I strongly agree with the majority view that "Open" and "Save As" should behave like they do in other applications, I actually think that remembering the last location (@MarcSabatella's preferred behaviour) makes more sense for the "Export" dialog (and possibly for "Safe a Copy" depending on what other programs do).

Look at it this way:

- "Open" and "Save As" affect the currently displayed score, so it makes sense for these dialogs to default to the location of the current score.
- "Export" and "Save a Copy" do not affect the current score: they create *new* scores. It makes sense for these dialogs to default to the location last exported/copied to.

**Comment by anatoly-os** [comment]
IMHO, if some people like unusual behaviour, it doesn't mean it is beloved by vast majority, so we shouldn't keep it. Users have their personal experience with other software which behaves the same in the case of Save As... If MuseScore is different it complicates user scenarios and becomes one more obstacle on the learning curve.

**Comment by MarcSabatella** [comment]
I'm not as option-averse as some, but I don't really see the need here.  To me it's no big deal either way even though I have a preference for the status quo.  FWIW, though, when this first came up a few years ago, I checked a bunch of applications and found were hardly alone in doing things this way, but it was indeed the minority of applications..

**Pull request merged**